### PR TITLE
Avoid LINQ allocations

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Part of https://github.com/dotnet/aspnetcore/issues/30247.

I identified these LINQ queries as causing ~5% of allocations in my testing so let's do the more verbose thing for efficiency.

### CODE REVIEW

**Review by NTaylorMullen — APPROVED** [review]

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs [hunk 1]
```diff
@@ -85,10 +85,19 @@ public TagHelperBinder(string tagHelperPrefix, IEnumerable<TagHelperDescriptor>
             Dictionary<TagHelperDescriptor, IReadOnlyList<TagMatchingRuleDescriptor>> applicableDescriptorMappings = null;
             foreach (var descriptor in descriptors)
             {
-                var applicableRules = descriptor.TagMatchingRules.Where(
-                    rule => TagHelperMatchingConventions.SatisfiesRule(tagNameWithoutPrefix, parentTagNameWithoutPrefix, attributes, rule));
+                // We're avoiding desccriptor.TagMatchingRules.Where and applicableRules.Any() to avoid
+                // Enumerator allocations on this hotpath
+                var applicableRules = new List<TagMatchingRuleDescriptor>();
```
Comment (sharwell) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs:90]: 💡 This should start as `null` and get a value assigned only when adding.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs [hunk 2]
```diff
@@ -85,10 +85,19 @@ public TagHelperBinder(string tagHelperPrefix, IEnumerable<TagHelperDescriptor>
...
+                    {
+                        applicableRules.Add(rule);
+                    }
+                }
 
-                if (applicableRules.Any())
+                if (applicableRules.Count > 0)
                 {
                     if (applicableDescriptorMappings == null)
                     {
```
Comment (sharwell) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs:103]: 💡 The `ToList()` call below this line creates a new copy of the list

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs [hunk 3]
```diff
@@ -85,10 +85,19 @@ public TagHelperBinder(string tagHelperPrefix, IEnumerable<TagHelperDescriptor>
             Dictionary<TagHelperDescriptor, IReadOnlyList<TagMatchingRuleDescriptor>> applicableDescriptorMappings = null;
             foreach (var descriptor in descriptors)
             {
-                var applicableRules = descriptor.TagMatchingRules.Where(
-                    rule => TagHelperMatchingConventions.SatisfiesRule(tagNameWithoutPrefix, parentTagNameWithoutPrefix, attributes, rule));
```
Comment (sharwell) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs:89]: 💡 We could actually keep this form if we do something like this:

```csharp
var applicableRules = descriptor.TagMatchingRules.WhereAsReadOnlyList(
    static (rule, arg) => TagHelperMatchingConventions.SatisfiesRule(arg.tagNameWithoutPrefix, arg.parentTagNameWithoutPrefix, arg.attributes, rule),
    (tagNameWithoutPrefix, parentTagNameWithoutPrefix, attributes));
```

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs [hunk 4]
```diff
@@ -87,24 +87,29 @@ public TagHelperBinder(string tagHelperPrefix, IEnumerable<TagHelperDescriptor>
...
+                List<TagMatchingRuleDescriptor> applicableRules = null;
                 for (var i = 0; i < descriptor.TagMatchingRules.Count; i++)
                 {
                     var rule = descriptor.TagMatchingRules[i];
                     if (TagHelperMatchingConventions.SatisfiesRule(tagNameWithoutPrefix, parentTagNameWithoutPrefix, attributes, rule))
                     {
+                        if (applicableRules is null)
+                        {
+                            applicableRules = new List<TagMatchingRuleDescriptor>();
+                        }
```
Comment (sharwell) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs:99]: 💡 

```csharp
applicableRules ??= new List<TagMatchingRuleDescriptor>();
```

**Pull request merged**